crates/wast/src/parser.rs

```diff
@@ -843,83 +1009,77 @@ impl<'a> Cursor<'a> {
-                }
-                if n == usize::MAX {
-                    return None;
-                }
-                let (token, _) = &self.parser.buf.tokens[n];
-                match token {
-                    Source::Token(t) => {
-                        self.cur = n + 1;
-                        Some(t)
-                    }
-                    _ => unreachable!(),
-                }
+            // Note that this entire operation happens in a loop (hence the
+            // "somewhat cached") because the set of known annotations is
+            // dynamic and we can't cache which annotations are skipped. What we
+            // can do though is cache the number of tokens in the annotation so
+            // we know how to skip ahead of it.
+            let mut n = next.get();
+            if n == 0 {
+                n = self.find_next().unwrap_or(usize::MAX);
+                next.set(n);
+            }
+            if n == usize::MAX {
+                return None;
             }
+            self.cur = n;
         }
     }
 
-    fn is_unknown_annotation_start(&self) -> bool {
+    fn annotation_start(&self) -> Option<&'a str> {
         match self.parser.buf.tokens.get(self.cur).map(|p| &p.0) {
             Some(Source::Token(Token::LParen(_))) => {}
-            _ => return false,
+            _ => return None,
         }
         let reserved = match self.parser.buf.tokens.get(self.cur + 1).map(|p| &p.0) {
             Some(Source::Token(Token::Reserved(n))) => n,
-            _ => return false,
+            _ => return None,
         };
-        let annotation = if reserved.starts_with("@") && reserved.len() > 1 {
-            &reserved[1..]
+        if reserved.starts_with("@") && reserved.len() > 1 {
+            Some(&reserved[1..])
         } else {
-            return false;
-        };
-
-        // TODO: need to make this a dynamic check
-        annotation != "custom" && annotation != "name"
+            None
+        }
     }
 
     /// Finds the next "real" token from the current position onwards.
     ///
     /// This is a somewhat expensive operation to call quite a lot, so it's
     /// cached in the token list. See the comment above in `advance_token` for
     /// how this works.
-    fn find_next(&mut self) -> Option<()> {
+    ///
+    /// Returns the index of the next relevant token to parse
+    fn find_next(mut self) -> Option<usize> {
+        // If we're pointing to the start of annotation we need to skip it
+        // in its entirety, so match the parentheses and figure out where
+        // the annotation ends.
+        if self.annotation_start().is_some() {
+            let mut depth = 1;
+            self.cur += 1;
+            while depth > 0 {
+                match &self.parser.buf.tokens.get(self.cur)?.0 {
+                    Source::Token(Token::LParen(_)) => depth += 1,
+                    Source::Token(Token::RParen(_)) => depth -= 1,
+                    _ => {}
```

Comment:
same thign as in the last review about strings and quote tokens

Reply:
ah nvm, I see that string literals are a whole token :ok_hand:

Reply:
Ah yes indeed, they're whole tokens so no need to worry about them here